siq ixdat integration

ScottSoren opened the pull request:
Hi Spectro Inlets,
This PR includes a few changes we have talked about (and maybe some minore ones we haven't) that make spectro_inlets_quantification (siq) ready for use with ixdat.

The accompanying version ixdat, with updates to be merged in a PR immediately following this one, is here: https://github.com/ixdat/ixdat/pull/116
Development scripts which I have used with (unpublished) data from the lab at ICL are here: https://github.com/ScottSoren/quant_integration

What was done:
1. Molecule data files are converted from json to yaml and the built-in molecule library is expanded to include those essential for aqueous EC-MS (H2, He, H2O, N2, O2, Ar, and CO2). Yaml is prefered because of its ability to include comments. The molecule data files are commented to indicate the source of the data. I suggest welcoming additions to the library by contributors, with the condition that all data sources be publically available.
2. Auxiliary data directory. When a Molecule is loaded, siq looks first, if available, an auxiliary library which is pointed to by `config.aux_molecule_directory`, and then in its internal library. `config.aux_molecule_directory` can be set by a user or third-party package such as ixdat. With ixdat, usage can look like:
```
import ixdat

ixdat.plugins.activate_si_quant()
ixdat.plugins.si_quant.QUANT_DIRECTORY = "~/projects/batteries/quantification_data"
```
The auxiliary data directory enables public or private researchers to maintain their own database with additional molecules or corrections to the publically available data in siq's internal library with data gathered in-house. User's calibration files will also go in this auxiliary data directory. 
3. The methods `SensitivityFactor.__add__`, `Calibration.scaled_by_factor`, and `Calibration.add_isotopes` have been added. This incorporates all of the code that had been in `ixdat.si_quant_patch` into siq. I have typed these methods to fit in with the style of siq, but this was a bit tricky for `__add__` since it is highly flexible to take `SensitivityFactor`, `SensitivityList`, `CalPoint`, or `Calibration` as input. It enables this, for example:
```
from spectro_inlets_quantification import CalPoint, SensitivityFactor, Calibration

CO2_isotopes_cal = Calibration(
    cal_list=[CalPoint(mol="CO2", mass="M44", F=1, setup="Ifan2")]
)
CO2_isotopes_cal.add_isotopes({"CO2": ("M44", ["M46", "M48"])})

calibration = (
    CalPoint(mol="O2", mass="M32", F=1.2, setup="Ifan2")
    + SensitivityFactor(mol="He", mass="M4", F=0.2)
    + CO2_isotopes_cal
)
calibration.plot_as_spectrum()
```
Of course, with si_quant activated in ixdat, the `Calibration`s and `CalPoint`s will typically be returned by methods of `MSMeasurement` and inheriting classes.
4. The package is now ready for uploading to pypi. Please check the `setup.py`. I couldn't tell what "DUAL_LICENSE" is on the [PyPI classifiers](https://pypi.org/classifiers/). It works: I uploaded a distribution to the PyPI test site: https://test.pypi.org/project/spectro-inlets-quantification/1.1/

Looking forward to merging this on the real pyPI so that EC-MS + ixdat users will have all these tools effortlessly at their fingertips.

Commit:
yaml molecules (not json), and aux data directory

Commit:
move from ixdat.si_quant_patch: calibration addition, scaling, isotopes

Commit:
update molecule tests: yml instead of json

Commit:
get ready for pypi

KennethNielsen commented on setup.py:
These changes really should go in the pyproject.toml, which I have opted into for this sort of project metadata for this projects, but that is my mistake for leaving the setup.py in place. I will see about porting it later.

KennethNielsen commented on src/spectro_inlets_quantification/calibration.py:
```diff
@@ -139,7 +139,7 @@ class Calibration(SensitivityList):
         self,
         # calibration basics:
-        cal_list: Optional[List[CalPoint]] = None,
+        cal_list: Optional[List[SensitivityFactor]] = None,
```
Ok, so we want to allow providing SensitivityFactor here, instead of only the more specific CalPoint. Can you modify the class docstring and the args list in the init docstring to reflect this?

KennethNielsen commented on src/spectro_inlets_quantification/calibration.py:
```diff
@@ -680,3 +728,20 @@ def predict_F(self, mol: MOL, mass: MASS) -> float:
+
+
+def scale_by_factor(calibration: Calibration, factor: float) -> Calibration:
+    """Return copy of `calibration` w all sensitivity factors multiplied by `factor`"""
```
with or w.

KennethNielsen commented on src/spectro_inlets_quantification/calibration.py:
```diff
@@ -680,3 +728,20 @@ def predict_F(self, mol: MOL, mass: MASS) -> float:
+
+def scale_by_factor(calibration: Calibration, factor: float) -> Calibration:
+    """Return copy of `calibration` w all sensitivity factors multiplied by `factor`"""
+    from spectro_inlets_quantification import CalPoint, Calibration
```
This seems left over. These classes are in this module.

KennethNielsen commented on src/spectro_inlets_quantification/calibration.py:
```diff
@@ -680,3 +728,20 @@ def predict_F(self, mol: MOL, mass: MASS) -> float:
+    new_calibration_as_dict = calibration.as_dict().copy()
+
+    cal_list = []
+    for cal in calibration.cal_list:
```
I'd prefer to not abbreviate where we don't have to, so let's go with calibration

ScottSoren replied:
`cal` for a `CalPoint` is a common and (I think) unproblematic abbreviation in siq, and writing out `calibration_point` would force at least one of the following lines to be broken up. So I prefer to keep the abreviation here.

KennethNielsen commented on src/spectro_inlets_quantification/calibration.py:
```diff
@@ -680,3 +728,20 @@ def predict_F(self, mol: MOL, mass: MASS) -> float:
+        new_cal_as_dict = cal.as_dict()
+        new_cal_as_dict.update(F=cal.F * factor, F_type=cal.F_type + " corrected")
+        cal_list.append(CalPoint(**new_cal_as_dict))
+    new_calibration_as_dict.update(cal_list=cal_list)
```
`new_calibration_as_dict["cal_list"] = cal_list` when there is only one item being changed

KennethNielsen commented on src/spectro_inlets_quantification/molecule.py:
```diff
@@ -178,42 +178,53 @@ def norm_spectrum(self) -> MASS_TO_FLOAT:
-        with open(path_to_json) as json_file:
-            self_as_dict = json.load(json_file)
+        self_as_dict = {}
+        for mol_dir in [CONFIG.molecule_directory, CONFIG.aux_molecule_directory]:
```
I wonder. Should the order be the other way around, so that it will be possibly to override the built any builtins with an identicially named version of their own?

ScottSoren replied:
Yes! :) 
Fixed. The order is implemented, as you suggest in the lead comment, in `CONFIG.data_directories`

KennethNielsen commented on src/spectro_inlets_quantification/molecule.py:
```diff
@@ -178,42 +178,53 @@ def norm_spectrum(self) -> MASS_TO_FLOAT:
+
         self_as_dict.update(kwargs)
-        # Unfortunately, saving and loading a dict with integer keys to json
+        # Unfortunately, saving and loading a dict with integer keys to yaml
```
This is true of yaml as well?

ScottSoren replied:
No! Beautiful. Yaml can deal with integer dict keys. Code and data files simplified accordingly :)

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -350,7 +353,7 @@ def filter(self, **kwargs: SENSITIVITYLIST_FILTER_TYPE) -> "SensitivityList":  #
     def as_dict(self) -> Dict[str, List[SENSITIVITYFACTOR_AS_DICT]]:
         """Return a full dictionary representation of the sensitivity list."""
-        sf_dicts = [sf.as_dict for sf in self.sf_list]
+        sf_dicts = [sf.as_dict() for sf in self.sf_list]
```
whoops

ScottSoren replied:
I found this due to typing! :)

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -1137,3 +1140,59 @@ def plot_F_vs_f(
+
+
+def append_sensitivity_factors(*args):
+    """Recursively append elements in a list"""
```
As much as I like re recursive (reductive) approach to this, I'm wondering if a iterative approach might not be easier to read (the other stuff isn't done that much anymore). You can also use a unpackagin/packing trick to split the list up:
```python
first, second, *remaining = args
result = append_two(first, second)
for sf in remaining:
    result  = append_two(result, sf)
return result
```

You could even unpack only the first and let that be result, and then the loop, so you will not need the extra add for the first two, but then you will different behavior for the special case of 1 argument.

KennethNielsen replied:
But I don't mind it as is, so view it as an alternative suggestion.

ScottSoren replied:
I agree the unpacking trick is a bit simpler and much easier to read! Will use it next time - this code is not needed when it is easy to just use `+`.

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -1137,3 +1140,59 @@ def plot_F_vs_f(
+
+
+def append_two_sensitivity_factors(
+        sf_1: Union[SensitivityFactor, SensitivityList],
```
If you find this sort of type definition unhelpful for the reading experience, you can consider defining a TypeAlias by which you give a custom type, like a this case a Union, a name of its own. In this case it could be something like "SensitivityLike" or similar, to indicate that the type is is something that behaves like something or can be used as something. You can see an example of this in the types for gases, where I made the MixtureLike, which I think is an alias for the string/dict/Miuxture that a lot of methods accept.

ScottSoren replied:
I find it fine for the reading process, but agree with what we talked about with a SensitivityLike base class... but I didn't understand the implementation difference between that and a mix-in class, so I didn't touch it for now.

KennethNielsen replied:
I think the two discussions are separate. This comment was **only** about the type definitions.

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -1137,3 +1140,59 @@ def plot_F_vs_f(
+        sf_2 (SensitivityFactor or SensitivityList): The second sensitivity factor or
+            list thereof
+
+    Returns SensitivityList: a `SensitivityList` containing all of the sensitivity
```
SensitivityList: A ...

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -1137,3 +1140,59 @@ def plot_F_vs_f(
+        factors in `sf_1` and `sf_2`. This will be a `Calibration` unless both `sf_1`
+        and `sf_2` are both themselves plain `SensitivityList`s
+    """
+    from .calibration import Calibration
```
Should no longer be necessary

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -1137,3 +1140,59 @@ def plot_F_vs_f(
+            return sf_1 + sf_2  # this is implemented in spectro_inlets_quantification
+        elif isinstance(sf_2, SensitivityFactor):
+            sf_list = sf_1.sf_list + [sf_2]
+            obj_as_dict: Any = sf_1.as_dict()
```
There actually is a type for this, see `SensitivityFactor.as_dict`

ScottSoren replied:
Yeah, but that type wasn't working for me... I couldn't figure out why not.

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -1137,3 +1140,59 @@ def plot_F_vs_f(
+
+    if isinstance(sf_1, SensitivityList):
+        if isinstance(sf_2, SensitivityList):
+            return sf_1 + sf_2  # this is implemented in spectro_inlets_quantification
```
This comment should probably be reworded or removed

ScottSoren replied:
this function has been deleted for an implementation directly in the respective `__add__` methods.

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -1137,3 +1140,59 @@ def plot_F_vs_f(
+            sf_list = [sf_1, sf_2]
+            return Calibration(cal_list=sf_list)
+        else:
+            return append_sensitivity_factors(sf_2, sf_1)
```
Isn't this an infinite recursion?

ScottSoren replied:
This code is gone now either way.

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -1137,3 +1140,59 @@ def plot_F_vs_f(
+        sf_1: Union[SensitivityFactor, SensitivityList],
+        sf_2: Union[SensitivityFactor, SensitivityList]
+) -> SensitivityList:
+    """This should be incorporated into __add__ of SensitivityFactor and SensitivityList
```
So, while I appreciate that this may be more that you bargained for in this PR, I think it would be nice if this code was in fact merged into those classes. The thing is that the whole outcome tree below is fairly complex. Merging it into the classes would eliminate some of that complexity because one of the types would be known.

KennethNielsen reviewed (changes requested):
This is great. There are a bunch of nitpick in different places, but mostly small stuff. There are however two larger comments:

1) As indicated inline, I think it would be great to have the add functionality in the functions incorporated into the classes and mentioned in the docstring. I know it is more work, but I think it will make it easier to reason about, when the tree of potential outcomes when you add stuff is less complex, because one of the types in known in advance.

2) With regards to the aux folder stuff, this is implemented a little differently than I was expecting and perhaps also from the functionality I would expect it to add. The central point is that I considered the "aux" folder to be a layer on top of the builtin one, meaning that if a molecule (or anything else) is only present in one of the two folders, you will of course use the one that is there (like it is now) **but** if they are both there (i.e. a "He.yml" in both folders) then it would use the aux one, this allowing the user to override the builtin data files, if they think there is an error and can't wait for a release, or if they want to expand it. Furthermore, in the case where both normal data dir and aux data dir is present I would expect the default save location to be the aux dir, since saving to the builtin folder of a distributed package would be weird. The way I imagined this being done, would be to expand the class to be saved / loaded or the config class, with functionality to built the correct path for loading/saving. In the Config class it might look like this:

```python
class Config(metaclass=Singleton):
    ...

    def __init__(self):
        ...
        self.data_directories = [self._data_directory]

    @aux_data_directory.setter
    def aux_data_directory(self, path: Path) -> None:
        self._aux_data_directory = Path(path)
        self.data_directories = [self._aux_data_directory, self._data_directory]

    def get_molecule_path(molecule_name: str):
        """Return the molecule load/save path for ``molecule_name``"""
        for data_dir in self.data_directories:
            if data_dir is not None:
                return data_dir / "molecules" / (molecule_name + ".yml")
        raise RuntimeError("`data_directories` have no paths")
```

This would also have the effect of making the changes in the e.g. the Molecule class (and the other) very small.

ScottSoren committed:
implement review: SensitivityList.__add__ code in place.

ScottSoren commented:
Hey @KennethNielsen , thanks for the review! 
I have implemented the requested changes, and it looks much better with the `__add__` code where it belongs :) 
I also improved the implementation of the auxiliary directory. But one thing that is kind of half-done is the auxiliary directories for the other stuff (chips, calibrations, signal processors). With the implementation for molecules it would require copying around that kind of ugly loop with a try/except, so I think some code path could be shared. 

They should also probably be yml instead of json. But it would It works fine as is with the example scripts (they save a calibration by specifying the exact path within the user's project). So Leave it for another PR?

ScottSoren commented:
CLA signed as per the dual license agreement.

KennethNielsen commented:
> CLA signed as per the dual license agreement.

Thanks

KennethNielsen commented:
> Hey @KennethNielsen , thanks for the review! I have implemented the requested changes, and it looks much better with the `__add__` code where it belongs :) I also improved the implementation of the auxiliary directory. But one thing that is kind of half-done is the auxiliary directories for the other stuff (chips, calibrations, signal processors). With the implementation for molecules it would require copying around that kind of ugly loop with a try/except, so I think some code path could be shared.

Cool

> They should also probably be yml instead of json. But it would It works fine as is with the example scripts (they save a calibration by specifying the exact path within the user's project). So Leave it for another PR?

Yes, that's fine

KennethNielsen commented on src/spectro_inlets_quantification/calibration.py:
```diff
@@ -273,7 +273,7 @@ def load(cls, file_name: PATHLIKE, cal_dir: PATHLIKE = None, **kwargs: Any) -> "
             Calibration: A calibration object ready to quantify your data!
         """
-        cal_dir = cal_dir or CONFIG.calibration_directory
+        cal_dir = cal_dir or CONFIG.calibration_directories[0]
```
This seems to only draw from folder one, but we had talked about the ability to seamlessly mix, providing overrides for the ones that are not supplied and using defaults for the rest, where it makes sense. I have a suggestion for this, I'll write it up in a PR against this branch.

ScottSoren replied:
your PR is merged.

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -133,6 +133,18 @@ def copy(self) -> "SensitivityFactor":
+            return other + self
+        elif isinstance(other, SensitivityFactor):
+            return SensitivityList(sf_list=[self, other])
+        raise TypeError(f"can't add {self} and {other} together")
```
Nice

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -218,13 +230,41 @@ def __len__(self) -> int:
 
-    def __add__(self, other: "SensitivityList") -> "SensitivityList":
-        """Return a `SensitivityList` with the sensitivity factors of itself and ``other``."""
+    def __add__(self, other: [SensitivityFactor or "SensitivityList"]) -> "SensitivityList":
```
I was not aware that you could write the types like that. Are you sure it does what you want it to? Since SensitivityFactor is a normal class I would expect this to return just [SensitivityFactor], which I don't think does what you want. The canonical way I know of would be `Union[SensitivityFactor, "SensitivityList"]`

ScottSoren replied:
No, I was not sure. This looks like mistake on my part. Switched to the cononical way.

KennethNielsen commented on src/spectro_inlets_quantification/sensitivity.py:
```diff
@@ -218,13 +230,41 @@ def __len__(self) -> int:
-        else:
-            raise TypeError(f"can't add {other} to {self}")
-        return SensitivityList(sf_list)
+            cls = self.__class__
```
Maybe a bit of nitpicking her, but here you give the class an alias, above you just use it from `self.__class__`, it seems a little random

ScottSoren replied:
Now the alias `cls` is used both places. I like the habit of using an alias for this because sometimes the determination of the type to return is more involved.

KennethNielsen reviewed (changes requested):
Looks great. I'll mark this as "Requests changes" and will immediately submit a PR against your branch with an alternative suggestion for the directory handling.

Commit:
Unify override, aux  base data file look-up logic

Commit:
Add tests for data file lookup code

ScottSoren committed:
Merge pull request #1 from KennethNielsen/level_config_dirs

Level config dirs

ScottSoren committed:
update docstring to get_best_data_file

ScottSoren committed:
respond to #1 review

ScottSoren committed:
finish converting json to yaml

ScottSoren committed:
improve object saving with get_save_destination

ScottSoren committed:
include data files when packaging, fix pyproject.toml

ScottSoren commented:
Hey, @KennethNielsen , this is ready for you. 
I've merged your improvement of the best file selection, and implemented it for saving as well.
I've also switched json out for yaml in the other saveables.
Finally, I tinkered a bit with setup.py and pyproject.toml to get it packaging correctly. 

I've released it on the test PyPI and now we are using this at ICL:
https://test.pypi.org/project/spectro-inlets-quantification/1.2.1/

KennethNielsen commented on pyproject.toml:
```diff
@@ -20,11 +20,12 @@ classifiers = [
-    "matplotlib>=3.7.1",
-    "pandas>=2.0.2",
-    "attrs>=23.1.0",
+    # please don't increase these version numbers without a reason.
```
Commonly they would be suggested to be increased by dependabot, due to critical or security bugs. Is this packaged being used in a setting, where packages cannot easily be upgraded. If "yes" I would like to know which settings, so I know what forms the limit for upgrading.

ScottSoren replied:
I just know that pip could not auto-upgrade the packages to the required versions, preventing me from installing it on my own computer... and that it works fine with the versions that I have.

I upgraded pip and it still doesn't have access to those versions of scipy and numpy. Perhaps I would have to upgrade python itself for the higher versions (I am on python 3.9.13).

KennethNielsen replied:
This is puzzling. Could you try with a fresh environment. The tests run (and install from that requirement list) in CI on python versions 3.8 -> 3.11, so I really don't see any reason why should be a problem.

ScottSoren replied:
Works now, and I've reverted the version numbers! See top-level comment below.

KennethNielsen commented on pyproject.toml:
```diff
@@ -40,7 +41,8 @@ dev = [
     "myst-parser",
     "sphinx-rtd-theme",
-    "git+https://github.com/executablebooks/sphinx-proof@d37495f",
+    # This results in an error when running setup.py:
```
That I will have to look into. Last time I checked, we cannot build the docs wihtout this specific version, which contains / contained unreleased features. But also, the development dependencies should not be involved in the package. I will have to look into that.

KennethNielsen replied:
So, I checked. There is no new release of this package yet, so we still depend on it. Please revert this change, and I will take a look into the packaging.

KennethNielsen reviewed (changes requested):
Looks good to me. Only thing left is a question about in which setting newer package version would be a problem and then we should merge.

ScottSoren commented:
@KennethNielsen , I got it working with these nice new version numbers.

Two things I had to realize:

- scipy has a maximum numpy version. So scipy has to be upgraded before numpy. When I do that manually it works. Perhaps the order should be switched in requirements.txt?
- numpy by default requires administrator access on Windows. This behavior can be changed by adding the --user tag:
    `pip install --upgrade numpy --user`,
    but I don't think this tag passes through to the requiremets.txt when installing siqms, so it has to be done manually.

It makes sense that these don't cause a problem in a clean environment, but those at ICL using this package installed Anaconda at some unknown time in the past and don't want to worry about environments. Now that I know, I can tell them, but maybe the above points should be stated in documentation somewhere as well? (Same applies for ixdat documentation).

ScottSoren committed:
revert requirement version numbers; fix typo in He

ScottSoren committed:
fix calibration addition infinite recursion bug

KennethNielsen commented:
Good point about installation instructions. However, I do think that there is some more stuff we need to dive into before we write those instructions. Like e.g. if you are using anaconda, the user should use `conda` to upgrade all packages before install, not `pip`. With regards to installation on Windows requiring higher priviledges and order of packages mattering, I have never experiences this when working with venvs and in fact many sort their requirements alphabetically, so maybe I missed it or your env is a little broken. I'd like to sort that out before making more difficult installation instructions.

With regards to how new version of packages we require etc. We could throttle down to some older state. But I want a mechanism for determining what is acceptable. Like e.g. "available in anaconda release as of two years ago" or something like that.

If I can convince you to revert the deletion of the sphinx proof requirement I will merge it and then I will give the packaging an iteration myself.

ScottSoren committed:
revert commenting out of sphinx requirement

ScottSoren commented:
Sounds good! I've reverted the sphinx requirement to leave to you.

I guess a full installation guide might be out of scope since different people run python differently. But most users known to me (including myself) are not working with venvs (and my anaconda python environment is not broken). Both the order of updating and the --user flag requirements were experienced by several at Imperial who had installed Anaconda ~two years earlier. Conda installing may have been the right solution, but we didn't think of it since we were using pip for siq and ixdat.

That reminds you, since you're packaging, a suggestion:
`spectro_inlets_quantification` is very long to type. `siq` and `siquant` are both taken on PyPI. What about `siqms`? siqms is a straight-forward abbreviation of Spectro Inlets Quantitative Mass Spectrometry, and easy to pronounce ("sikems"). 
Obviously the name decision lies with SI :)
